src/Temporalio/Worker/WorkflowWorker.cs

```diff
@@ -0,0 +1,254 @@
+        /// </summary>
+        internal TemporalWorker Worker { get; private init; }
+
+        /// <summary>
+        /// Gets the known set of workflow definitions by name.
+        /// </summary>
+        internal Dictionary<string, WorkflowAttribute.Definition> WorkflowDefinitions { get; private init; }
+
+        /// <summary>
+        /// Execute this worker until poller shutdown or failure.
+        /// </summary>
+        /// <returns>Task that only completes successfully on poller shutdown.</returns>
+        public async Task ExecuteAsync()
+        {
+            using (logger.BeginScope(new Dictionary<string, object>()
+            {
+                ["TaskQueue"] = Worker.Options.TaskQueue!,
+            }))
+            {
+                var tasks = new ConcurrentDictionary<Task, bool?>();
+                var codec = Worker.Client.Options.DataConverter.PayloadCodec;
+                while (true)
+                {
+                    var act = await Worker.BridgeWorker.PollWorkflowActivationAsync().ConfigureAwait(false);
+                    // If it's null, we're done
+                    if (act == null)
+                    {
+                        break;
+                    }
+                    // Otherwise handle and put on task set
+                    // TODO(cretz): Any reason for users to need to customize factory here?
+                    var task = Task.Run(() => HandleActivationAsync(codec, act)).
+                        ContinueWith(task => { tasks.TryRemove(task, out _); }, TaskScheduler.Current);
+                    tasks[task] = null;
+                }
+                // Wait on every task to complete (exceptions should never happen)
+                await Task.WhenAll(tasks.Keys).ConfigureAwait(false);
+                var deadlockCount = deadlockedWorkflows.Count;
+                if (deadlockCount > 0)
+                {
+                    logger.LogWarning(
+                        "Worker shutdown with {DeadlockedTaskCount} deadlocked workflow tasks still running",
+                        deadlockCount);
+                }
+            }
+        }
+
+        private async Task HandleActivationAsync(IPayloadCodec? codec, WorkflowActivation act)
+        {
+            WorkflowActivationCompletion comp;
+            RemoveFromCache? removeJob = null;
+
+            // Catch any exception as a completion failure
+            try
+            {
+                // Decode the activation if there is a codec
+                if (codec != null)
+                {
+                    await WorkflowCodecHelper.DecodeAsync(codec, act).ConfigureAwait(false);
+                }
+
+                // Log proto at trace level
+                logger.LogTrace("Received workflow activation: {Activation}", act);
+
+                // We only have to run if there are any non-remove jobs
+                removeJob = act.Jobs.Select(job => job.RemoveFromCache).FirstOrDefault(job => job != null);
```

Comment:
@Sushisource can / does core ensure that remove from cache would always be last?

Reply:
When would lang get remove from cache + other jobs?

Reply:
It's always last in the job list, but you can get other jobs if the WFT is about to be replied to and the cache is at the limit. So we finish the task, reply, then evict to make room for someone else

Reply:
I don't need to make that assumption I don't think unless a microbenchmark proves otherwise.

Reply:
Agree it doesn't matter for perf reasons, this was mostly for my understanding of the guarantees.